notebooks/NonClassicalSearch.ipynb

```diff
@@ -0,0 +1,1027 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's try to solve the same problem using AND-OR Graph Search."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 69,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Optional[[Action[name=Suck], if {A=Clean, B=Dirty, Loc1=A} then [Action[name=Right], Action[name=Suck]], if {A=Clean, B=Clean, Loc1=A} then []]]\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "null"
+      ]
+     },
+     "execution_count": 69,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import aima.core.agent.Action;\n",
+    "import aima.core.environment.vacuum.*;\n",
+    "import aima.core.search.agent.NondeterministicSearchAgent;\n",
+    "import aima.core.search.nondeterministic.AndOrSearch;\n",
+    "import aima.core.search.nondeterministic.NondeterministicProblem;\n",
+    "import static aima.core.environment.vacuum.VacuumEnvironment.*;\n",
+    "\n",
+    "NondeterministicSearchAgent<VacuumEnvironmentState, Action> agent = new NondeterministicSearchAgent<>(percept -> (VacuumEnvironmentState) percept);\n",
+    "\n",
+    "NondeterministicVacuumEnvironment world = new NondeterministicVacuumEnvironment(LocationState.Dirty, LocationState.Dirty);\n",
+    "world.addAgent(agent, LOCATION_A);\n",
+    "\n",
+    "NondeterministicProblem<VacuumEnvironmentState, Action> problem = new NondeterministicProblem<>(\n",
+    "        (VacuumEnvironmentState) world.getCurrentState(),\n",
+    "        VacuumWorldFunctions::getActions,\n",
+    "        VacuumWorldFunctions.createResultsFunction(agent),\n",
+    "        VacuumWorldFunctions::testGoal,\n",
+    "        (s, a, sPrimed) -> 1.0);\n",
+    "\n",
+    "AndOrSearch andOrSearch = new AndOrSearch();\n",
+    "System.out.println(andOrSearch.search(problem));"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Searching with Partial Observation\n",
+    "\n",
+    "Here the agent's percept does not suffice to pin down the exact state. Therefore, if the agent is one of the several possible states, then an action may lead to one of the several possible outcomes. The key concept required for solving the partially observable problems is the **belief state**, representing the agent's current belief state about the possible physical states it might be in, given the sequence of action and percept upto that point.\n",
+    "\n",
+    "When the agent's percept provide _no information at all_, we have what is called a **sensorless** problem. To solve sensorless problems, we search in the space of belief states rather than physical states. Notice that, in belief state space the problem is _fully observable_ because the agent always knows its own belief state. Furthermore, the solution(if any) is always a sequence of actions."
```

Comment:
We should add an example here where we search on the space of possible belief states.